opa/pkg/main.go

```diff
@@ -0,0 +1,263 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/types"
+	meta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/restmapper"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const outputFile = "/output/opa.json"
+
+func main() {
+	ctx := context.TODO()
+	instanceGvr := schema.GroupVersionResource{Group: "insights.fairwinds.com", Version: "v1", Resource: "customcheckinstances"}
+	checkGvr := schema.GroupVersionResource{Group: "insights.fairwinds.com", Version: "v1", Resource: "customchecks"}
+	config, err := ctrl.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+	dynamicInterface, err := dynamic.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+	kube, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err)
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(kube.Discovery())
+	if err != nil {
+		panic(err)
+	}
+	restMapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+	// TODO filter by namespace
+	checkInstances, err := dynamicInterface.Resource(instanceGvr).Namespace("").List(ctx, metav1.ListOptions{})
+
+	if err != nil {
+		panic(err)
+	}
+	actionItems := make([]ActionItem, 0)
+
+	for _, checkInstance := range checkInstances.Items {
+		var checkInstanceObject customCheckInstance
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(checkInstance.Object, &checkInstanceObject)
+		if err != nil {
+			panic(err)
```

Comment:
we should avoid panicking inside this loop, so we don't lose the whole report if there's a small issue

Reply:
Optionally, when images fail to scan in Trivy, we send that back to the server, and create a single action item saying `12 images failed to scan`. Maybe helpful here if e.g. we don't have the right RBAC to access a resource they asked us to check

Reply:
This is the one piece of feedback here I'm struggling with. Let's just think through this. Errors are either temporary or permanent. If it's a temporary error then failing the report seems appropriate, we don't end up with any flop in Action items, they get their failure notification (assuming they didn't turn it off) and odds are in an hour everything will be fine.

If it's a permanent error (like not having the right RBAC or bad Rego). Have an action item to fix it could make sense. But it's also directly user error.

So if we fail the report, it punishes the user for their mistake and they're more likely to get it fixed right away.
If we create an action item for the failure, then we're being gracious towards the user and giving them as many results as possible.

Let me try to look through the panics and see which ones feel like we should be creating action items for them. I think at least some of these will end up that way.

Reply:
OK yeah I see your point here.

I'd say, if it's something that would break upon the first installation, or upon creating a new policy, it's safe to `panic`. If it's something that could break because of a change in cluster state (e.g. I added a new Widget, which triggered this check for the first time), we should create an action item. Does that sound like a reasonable heuristic?